src/combinesummaries.1
.\" ==========================================================================
.\"         _   _      _   ____            __ __  __      _
.\"        | \ | | ___| |_|  _ \ ___ _ __ / _|  \/  | ___| |_ ___ _ __
.\"        |  \| |/ _ \ __| |_) / _ \ '__| |_| |\/| |/ _ \ __/ _ \ '__|
.\"        | |\  |  __/ |_|  __/  __/ |  |  _| |  | |  __/ ||  __/ |
.\"        |_| \_|\___|\__|_|   \___|_|  |_| |_|  |_|\___|\__\___|_|
.\"
.\"                  NetPerfMeter -- Network Performance Meter
.\"                 Copyright (C) 2009-2026 by Thomas Dreibholz
.\" ==========================================================================
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\" Contact:  dreibh@simula.no
.\" Homepage: https://www.nntb.no/~dreibh/netperfmeter/
.\"
.\" ###### Setup ############################################################
.Dd May 28, 2026
.Dt combinesummaries 1
.Os CombineSummaries
.\" ###### Name #############################################################
.Sh NAME
.Nm combinesummaries
.Nd Scalar Summary Combination Tool
.\" ###### Synopsis #########################################################
.Sh SYNOPSIS
.Nm combinesummaries Ar output\_file
.Op Ar variable\_names
.br
.Op Fl c Ar level | Fl \-compress Ar level
.br
.Op Fl s Ar separator | Fl \-separator Ar separator
.br
.Op Fl l | Fl \-line\-numbers | Fl n | Fl \-no\-line\-numbers
.br
.Op Fl q | Fl \-quiet
.Nm combinesummaries
.Op Fl h | Fl \-help
.Nm combinesummaries
.Op Fl v | Fl \-version
.\" ###### Description ######################################################
.Sh DESCRIPTION
.Nm combinesummaries
is a combination tool for CSV/GNU R data files: it creates a single data table from multiple input files. CombineSummaries supports on\-the\-fly BZip2 compression. After startup, the program accepts the following commands from standard input (*not* as command\-line arguments!):
.Bl -tag -width ident
.It Fl \-varnames=variables
A space\-separated list of output variable names to be added to the output data tables.
.It Fl \-values=values
Sets the values of the output variables for the next input file to the provided space\-separated values. One value for each variable has to be provided.
.It Fl \-input=filename
Process the given input file.
.It Fl \-simulationsdirectory=directory
Sets the directory where to find the input files.
.El
.Pp
.\" ###### Arguments ########################################################
.Sh ARGUMENTS
The following arguments may be provided:
.Bl -tag -width indent
.It Ar output\_file
The name of the output file to be created.
.It Op Ar variable\_names
A space\-separated list of output variable names to be added to the output data tables.
.It Fl c Ar level | Fl \-compress Ar level
Sets the BZip2 compression level; 1=none, 9=highest (default).
.It Fl s Ar separator | Fl \-separator Ar separator
Sets the separator for the tables. Default: tabulator character (i.e. $'\t').
.It Fl l | Fl \-line\-numbers
Generate line numbers for output rows (GNU R: row names). In this case, the input files must contain line numbers (GNU R: row names) as well! The original line numbers are kept in column SubLineNo. Default: off.
.It Fl n | Fl \-no\-line\-numbers
Do not add line numbers into the output table files, as usual for CSV (default).
.It Fl q | Fl \-quiet
Do not print verbose status information.
.It Fl h | Fl \-help
Prints command help.
.It Fl v | Fl \-version
Prints program version.
.El
.\" ###### Exit status ######################################################
.Sh EXIT STATUS
The
.Nm
tool exits with 0 on success, and >0 in case of an error.
.\" ###### Examples #########################################################
.Sh EXAMPLE
Consider two data tables:
.Pa v1.data.bz2
(created with ParameterY=Alpha) and
.Pa v2.data.bz2
(created with ParameterY=Beta).
The output data table should be written to
.Pa output.data.bz2 .
.br
( echo \-e "\-\-values=\\"Alpha\\"\\t\\"Test 1\\"" && \\
  echo "\-\-input=v1.data.bz2"               && \\
  echo \-e "\-\-values=\\"Beta\\"\\t\\"Test 2\\""  && \\
  echo "\-\-input=v2.data.bz2" ) | combinesummaries output.data.bz2 $'ParameterX\\tParameterY'
.\" ###### See also #########################################################
.Sh SEE ALSO
.Xr createsummary 1 ,
.Xr extractvectors 1
.\" ###### Notes ############################################################
.Sh NOTES
This program is part of NetPerfMeter. The latest version of NetPerfMeter can be found on the NetPerfMeter Homepage at
.Lk https://www.nntb.no/\(tidreibh/netperfmeter/ "NetPerfMeter Homepage" .
.\" ###### Authors ##########################################################
.Sh AUTHORS
Thomas Dreibholz,
.Lk https://www.nntb.no/\(tidreibh "Homepage"


ChangeLog
commit 1f383a60d2d770535581e46f78444bcf9c4f4ebf
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri May 29 11:29:02 2026 +0200

    * Make inclusion of examples (scripts, results) optional in CMakeLists.txt.
    * Moved obsolete test program.

commit d26562086455bbf41d21f6c3cae99b6cae1bf2a4
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri May 29 10:43:28 2026 +0200

    Improved FreeBSD port.

commit 946bfafbadddcea276a7aa360c8ee697695c69db
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri May 29 10:42:01 2026 +0200

    Fixed includes.

commit 9d186d0746f498d32427611d54b5a927e5baa773
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri May 29 09:58:58 2026 +0200

    RPM spec fix.

commit 26e1ca4d015ec9403a22df90b8600605839e44e7
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 16:13:55 2026 +0200

    New release netperfmeter-2.0.4.

commit 02576769d8562e11e6b0cb7d15200fe7d40e86cf
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 16:12:54 2026 +0200

    Minor manpage update.

commit bf272b25bfa5094a78b4657f4bd9912dcb23edb2
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 16:03:05 2026 +0200

    Minor clean-up.

commit b732393412bd54a0e4d0f1347ede6747b7a7a870
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 16:01:02 2026 +0200

    Added option to configure the separator.

commit 977df2a565b3cf11fba2e83e8ad00f11c8c570ec
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 15:53:07 2026 +0200

    Small clean-ups.

commit 862fb32f3d7abcac2fff6537fba36b34981d9ffa
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 15:35:21 2026 +0200

    Further clean-ups.

commit 1d6d37a6dfac84a1858845f33fd9df66ff1d4806
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 15:31:15 2026 +0200

    Further clean-ups.

commit 793d87f6b1d0746c09488178c0e84424ce544115
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 15:30:39 2026 +0200

    Added option to configure the separator.

commit 2148af89133efdc3beeffc3f0c4a090e9cb75468
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 14:28:24 2026 +0200

    Improved shell completions.

commit 336be30bc58436d386d92183522a263cc30d1fd7
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 14:22:41 2026 +0200

    Improved vector counting.

commit 8a5797d64ac46d71c275374869167aa5170cec12
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 14:19:01 2026 +0200

    Further improvements.

commit eb7c8be6269e2ca1790a0ec099201b3575516cc9
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 13:33:00 2026 +0200

    Code improvements.

commit 05228c17745801c8910c7d06570f9ab804a20215
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 00:13:34 2026 +0200

    CMakeLists.txt update.

commit 1b427da57b4c7563f2418755c0439f4b8b7b076f
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 28 00:08:58 2026 +0200

    Added more shell completions.

commit b799cc0179b7f960c9ef4d1fb28c5a9b34794e43
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed May 27 23:51:22 2026 +0200

    Options handling improvements.

commit 69b965905364e4306d1a8ef6149aa879136fca84
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed May 27 23:45:50 2026 +0200

    Packaging updates.

commit 82e1980494ab0fc8731b52671794c3c52a736c6b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed May 27 23:44:06 2026 +0200

    Added shell completions.

commit 8181abb986847144955c1e0b6e8da2c6dc863125
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed May 27 23:17:37 2026 +0200

    Further manpage improvements.

commit 3037cb9b5c9417375a288df233949c2e2ce9431a
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed May 27 23:06:07 2026 +0200

    Improved manpages.

commit 428bdbaa9bccd79ed23d82f2a02157974ff70c51
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed May 27 23:05:57 2026 +0200

    Added version option.

commit 5eb240346cab65c110a07b63b0a2e4d748898b97
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed May 27 23:05:40 2026 +0200

    Added version option.

commit ef9d7f39317e1017354170936b4d5a359dce0f0e
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue May 26 22:31:03 2026 +0200

    Improved CMakeLists.txt.

commit 58524abe9716d4f1a4f092f5aa64faec49fce966
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue May 26 21:03:35 2026 +0200

    Further manpage improvements.

commit 8c2c8ef1307836b855719897b2bcdb4775c11abc
Merge: 68c78ee b2d981b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue May 26 20:54:04 2026 +0200

    Merge branch 'master' of github.com:dreibh/netperfmeter

commit b2d981b677bc1bce0f2a9772a7905dff05375d3e
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue May 26 16:46:28 2026 +0200

    Improved manpages.

commit dadcd3763d363a0a401d64439996f24eefe8780a
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon May 25 22:37:19 2026 +0200

    A quick compile test under MacOS. Some more work would be needed to work around missing SCTP.

commit 68c78ee1c78dd0dc42e1131bae4064f039b1d7c3
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat May 9 10:36:06 2026 +0200

    Minor shell completion clean-up.

commit dcde13391c1f5dd16ce7f6db0da401713a3f310a
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue May 5 15:20:48 2026 +0200

    New release netperfmeter-2.0.3.

commit 3ee0511e03d1e0ebf4f642c984ceb802f4345e9d
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue May 5 15:04:13 2026 +0200

    Another portability improvement.

commit ae4fc20b2ab2caec15c235b6d3f0a9348dffb6e5
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue May 5 15:03:18 2026 +0200

    Portability improvements.

commit 78f2b7ddc69f2a4385b8e9cfcde7dcf9f26471ae
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue May 5 14:56:40 2026 +0200

    Clean-up of the includes.

commit 1d415efe3466e5eb4dd1d7019031c34555c577ce
Author: Thomas Dreibholz <dreibh@simula.no>

src/combinesummaries.cc
/*
 * ==========================================================================
 *         _   _      _   ____            __ __  __      _
 *        | \ | | ___| |_|  _ \ ___ _ __ / _|  \/  | ___| |_ ___ _ __
 *        |  \| |/ _ \ __| |_) / _ \ '__| |_| |\/| |/ _ \ __/ _ \ '__|
 *        | |\  |  __/ |_|  __/  __/ |  |  _| |  | |  __/ ||  __/ |
 *        |_| \_|\___|\__|_|   \___|_|  |_| |_|  |_|\___|\__\___|_|
 *
 *                  NetPerfMeter -- Network Performance Meter
 *                 Copyright (C) 2009-2026 by Thomas Dreibholz
 * ==========================================================================
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Contact:  dreibh@simula.no
 * Homepage: https://www.nntb.no/~dreibh/netperfmeter/
 */

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <cctype>
#include <getopt.h>
#include <iostream>
#include <unistd.h>

#include "inputfile.h"
#include "outputfile.h"
#include "package-version.h"


// ###### Check columns for proper quotation ################################
static bool checkColumns(const std::string& values)
{
   bool         inQuotation    = false;
   unsigned int quotationMarks = 0;
   for(size_t i = 0; i < values.size(); i++) {
      switch(values[i]) {
         case '\\':
            i++;
          break;
         case '"':
            inQuotation = !inQuotation;
            quotationMarks++;
          break;
      }
   }
   return ((quotationMarks & 1) == 0);
}


// ###### Read and process data file ########################################
void addDataFile(OutputFile&         outputFile,
                 const bool          withLineNumbers,
                 unsigned long long& outputLineNumber,
                 const std::string&  varNames,
                 const std::string&  varValues,
                 const std::string&  inputFileName,
                 const char*         separator = "\t")
{
   // ====== Open input file ================================================
   InputFile       inputFile;
   InputFileFormat inputFileFormat = IFF_Plain;
   if( (inputFileName.size() >= 4) &&
       ( (inputFileName.substr(inputFileName.size() - 4) == ".bz2") ||
         (inputFileName.substr(inputFileName.size() - 4) == ".BZ2")) ) {
       inputFileFormat = IFF_BZip2;
   }
   if(inputFile.initialize(inputFileName.c_str(), inputFileFormat) == false) {
      exit(1);
   }

   for(;;) {
      // ====== Read line from input file ===================================
      bool          eof;
      char          buffer[4097];
      const ssize_t bytesRead = inputFile.readLine(buffer, sizeof(buffer), eof);
      if((bytesRead < 0) || (eof)) {
         break;
      }

      // ====== Process line ================================================

      if(withLineNumbers) {
         if(inputFile.getLine() == 1) {
            if(outputLineNumber == 0) {
               if(outputFile.printf("%s%sSubLineNo%s%s\n", varNames.c_str(), separator, separator, buffer) == false) {
                  outputFile.finish();
                  exit(1);
               }
               outputLineNumber++;
            }
         }
         else {
            if(outputFile.printf("%llu%s%s%s%s\n",
                                 outputLineNumber, separator,
                                 varValues.c_str(), separator,
                                 buffer) == false) {
               outputFile.finish();
               exit(1);
            }
            outputLineNumber++;
         }
      }
      else {
         if(inputFile.getLine() == 1) {
            if(outputLineNumber == 0) {
               if(outputFile.printf("%s%s%s\n", varNames.c_str(), separator, buffer) == false) {
                  outputFile.finish();
                  exit(1);
               }
               outputLineNumber++;
            }
         }
         else {
            if(outputFile.printf("%s%s%s\n", varValues.c_str(), separator, buffer) == false) {
               outputFile.finish();
               exit(1);
            }
            outputLineNumber++;
         }
      }
   }

   inputFile.finish();
}


// ###### Version ###########################################################
[[ noreturn ]] static void version()
{
   std::cerr << "CombineSummaries" << " " << COMBINESUMMARIES_VERSION << "\n";
   exit(0);
}


// ###### Usage #############################################################
[[ noreturn ]] static void usage(const char* program, const int exitCode)
{
   std::cerr << "Usage:\n"
      << "* Run:\n  "
      << program << "\n"
         "    output_file\n"
         "    [variable_names]\n"
         "    [-c level|--compress level]\n"
         "    [-s separator|--separator separator]\n"
         "    [-l|--line-numbers|-n|--no-line-numbers]\n"
         "    [-q|--quiet]\n"
         "* Version:\n  " << program << " [-v|--version]\n"
         "* Help:\n  "    << program << " [-h|--help]\n";
   exit(exitCode);
}



// ###### Main program ######################################################
int main(int argc, char** argv)
{
   bool         quietMode        = false;
   bool         withLineNumbers  = false;
   unsigned int compressionLevel = 9;
   const char*  separator        = "\t";

   // ====== Handle command-line arguments ==================================
   const static struct option long_options[] = {
      { "compress",                  required_argument, 0, 'c' },
      { "separator",                 required_argument, 0, 's' },
      { "line-numbers",              no_argument,       0, 'l' },
      { "no-line-numbers",           no_argument,       0, 'n' },
      { "quiet",                     no_argument,       0, 'q' },

      { "help",                      no_argument,       0, 'h' },
      { "version",                   no_argument,       0, 'v' },
      {  nullptr,                    0,                 0, 0   }
   };

   int option;
   int longIndex;
   while( (option = getopt_long_only(argc, argv, "c:s:lnqhv", long_options, &longIndex)) != -1 ) {
      switch(option) {
         case 'c':
            compressionLevel = atol(optarg);
            if(compressionLevel < 1) {
               compressionLevel = 1;
            }
            else if(compressionLevel > 9) {
               compressionLevel = 9;
            }
          break;
         case 's':
            separator = optarg;
          break;
         case 'l':
            withLineNumbers = true;
          break;
         case 'n':
            withLineNumbers = false;
          break;
         case 'q':
            quietMode = true;
          break;
         case 'v':
            version();
          break;
         case 'h':
         case '?':
            // Exit with 0 on h/help, exit with 1 on '?' (unknown option):
            usage(argv[0], (option == 'h') ? 0 : 1);
          break;
         case '-':
          break;
         default:
            // This should not happen: wrong getopt parameters, or missing case?
            fprintf(stderr, "INTERNAL ERROR: Unhandled option c=%c code=%x!\n",
                    (isprint(option) ? (char)option : ' '), option);
            return 1;
          break;
      }
   }
   if(optind >= argc) {
      usage(argv[0], 1);
   }
   const std::string outputFileName(argv[optind++]);
   std::string       varNames;
   if(optind < argc) {
      varNames = argv[optind++];
      if(!checkColumns(varNames)) {
         std::cerr << "ERROR: Invalid quotation in variable names string!\n";
         exit(1);
      }
   }
   if(optind < argc) {
      std::cerr << "ERROR: Invalid option " << argv[optind] << "!\n";
      usage(argv[0], 1);
   }


   // ====== Print information ==============================================
   if(!quietMode) {
      std::cout << "CombineSummaries " << COMBINESUMMARIES_VERSION << "\n"
                << "* Compression Level: " << compressionLevel << "\n"
                << "* Separator:         \"" << separator << "\"\n"
                << "* Line Numbers:      " << (withLineNumbers ? "yes" : "no") << "\n"
                << "\n";
   }


   // ====== Open output file ===============================================
   OutputFile       outputFile;
   OutputFileFormat outputFileFormat = OFF_Plain;
   if( (outputFileName.size() >= 4) &&
       ( (outputFileName.substr(outputFileName.size() - 4) == ".bz2") ||
         (outputFileName.substr(outputFileName.size() - 4) == ".BZ2")) ) {
       outputFileFormat = OFF_BZip2;
   }
   if(outputFile.initialize(outputFileName.c_str(), outputFileFormat,
                            compressionLevel)== false) {
      exit(1);
   }


   // ====== Process input ==================================================
   unsigned long long outputLineNumber = 0;
   std::string        varValues;
   std::string        simulationsDirectory(".");
   char               commandBuffer[4097];
   char*              command;

   if(!quietMode) {
      std::cout << "Ready> ";
      std::cout.flush();
   }
   while((command = fgets(commandBuffer, sizeof(commandBuffer), stdin))) {
      size_t length = strlen(command);
      if( (length > 0) && (command[length - 1] == '\n') ) {
         command[length - 1] = 0x00;
         length--;
      }

      if(length == 0) {
         if(!quietMode) {
            std::cout << "Ready> ";
            std::cout.flush();
         }
         continue;
      }

      if(!(strncmp(command, "--values=", 9))) {
         varValues = &command[9];
         if( (varValues.size() >= 2)    &&
             (varValues.front() == '"') &&
             (varValues.back() == '"') ) {
            varValues = varValues.substr(1, varValues.size() - 2);
         }
         if(!checkColumns(varValues)) {
            std::cerr << "ERROR: Invalid quotation in variable values string!\n";
            exit(1);
         }
      }
      else if(!(strncmp(command, "--input=", 8))) {
         if(varValues.empty()) {
            std::cerr << "ERROR: No values given (parameter --values=...)!\n";
            exit(1);
         }
         addDataFile(outputFile, withLineNumbers, outputLineNumber,
                     varNames, varValues,
                     simulationsDirectory + "/" + &command[8],
                     separator);
         varValues = "";
      }
      else if(!(strncmp(command, "--varnames=", 11))) {
         varNames = &command[11];
         if( (varNames.size() >= 2)    &&
             (varNames.front() == '"') &&
             (varNames.back() == '"') ) {
            varNames = varNames.substr(1, varNames.size() - 2);
         }
         if(!checkColumns(varNames)) {
            std::cerr << "ERROR: Invalid quotation in variable names string!\n";
            exit(1);
         }
      }
      else if(!(strncmp(command, "--simulationsdirectory=", 23))) {
         simulationsDirectory = &command[23];
      }
      else {
         std::cerr << "ERROR: Invalid command!\n";
         exit(1);
      }

      if(!quietMode) {
         std::cout << "Ready> ";
         std::cout.flush();
      }
   }


   // ====== Close output file ==============================================
   unsigned long long in, out;
   if(!outputFile.finish(true, &in, &out)) {
      exit(1);
   }
   if(!quietMode) {
      std::cout << "\n" << "Wrote " << outputLineNumber << " lines";
      if(in > 0) {
         std::cout << " (" << in << " -> " << out << " - "
                   << ((double)out * 100.0 / in) << "%)";
      }
      std::cout << "\n";
   }

   return 0;
}
